db/hints: Use host ID to IP mappings to choose the ep manager to drain when node is leaving

dawmd opened the pull request:
In d0f5873, we introduced mappings IP–host ID between hint directories and the hint endpoint managers managing them. As a consequence, it may happen that one hint directory stores hints towards multiple nodes at the same time. If any of those nodes leaves the cluster, we should drain the hint directory. However, before these changes that doesn't happen – we only drain it when the node of the same host ID as the hint endpoint manager leaves the cluster.

This PR fixes that draining issue in the pre-host-ID-based hinted handoff. Now no matter which of the nodes corresponding to a hint directory leaves the cluster, the directory will be drained.

We also introduce error injections to be able to test that it indeed happens.

Fixes scylladb/scylladb#18761

piodul commented:
> This PR fixes the draining issues in pre-host-ID-based hinted handoff. Since one hint directory may correspond to multiple target nodes, if any of them leaves the cluster, draining should start. These changes ensure that. We also introduce error injections to be able to test that it indeed happens.

Consider the following situation:

- Have nodes A and B mapped to IP X,
- Node A leaves the cluster and draining starts,
- Write to replica A fails and we want to store a hint

What will happen in such case?

Wouldn't it be better to wait until _all_ nodes mapped to given IP leave the cluster?

dawmd commented:
> Consider the following situation:
> 
>     * Have nodes A and B mapped to IP X,
> 
>     * Node A leaves the cluster and draining starts,
> 
>     * Write to replica A fails and we want to store a hint
> 
> 
> What will happen in such case?

If the draining is still taking place, the hint will be rejected. If the draining has finished, a new hint endpoint manager will be created and the hint will be accepted.

> Wouldn't it be better to wait until _all_ nodes mapped to given IP leave the cluster?

It might be, yes. There's a trade-off here.

It would come with a risk of draining never happening. Let's assume we apply this logic and that we have a hint directory corresponding to a set of nodes `S`. Let's assume that all of the nodes but one – node `N` – have already left the cluster, and that `N` has host ID `H` and IP `I`. Let's also assume that there is a mapping `[H', I]`, where `H' != H`, i.e. `N` is related to the hint directory by IP only. Consider the following scenario:

1. `N` goes down and its IP is changed to `I'`, where `I != I'`
2. `N` is started again
3. `N` is decommissioned
4. No other node ever bears the IP `I` in the future

The hint directory will never be drained and when it comes to the migration of the hint directories, it will be removed.

On the other hand, performing the drain too early can make little sense if there's still a chance that the hints could be sent directly to the right node instead of applying them to all current replicas.

It shouldn't be too hard to change the logic, so if you think the alternative approach is better, let's go with it.

piodul commented:
> If the draining is still taking place, the hint will be rejected. If the draining has finished, a new hint endpoint manager will be created and the hint will be accepted.

Alright. As long as it happens only during the transition period then I think it's fine.

piodul commented:
The fix looks good to me. However, the PR description is too vague:

> This PR fixes the draining issues in pre-host-ID-based hinted handoff. Since one hint directory may correspond to multiple target nodes, if any of them leaves the cluster, draining should start. These changes ensure that.

You only wrote what the correct behavior should be, not what the current behavior is. Please include it in the PR description.

The first commit in the PR also seems to have this problem.

piodul commented:
Changing the backport label to 6.0, technically speaking this is a regression.

dawmd commented:
Version 2:
* Updated the description of the PR explaining the issues before the changes
* Updated the description of the PR explaining the issues before the changes

dawmd commented:
Version 3:
* one of the error injections introduced by this PR turned out to be redundant. It has been removed and the description of the commit adjusted. I also elaborated on what exactly we need the remaining error injection for

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9251/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9251/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9785/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1761/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9251/)
#### Failed Tests (784/30714):
- [test_counter_consistency_node_remove](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9251/testReport/junit/counter_tests/TestCountersOnMultipleNodes/Tests___dtest___test_counter_consistency_node_remove) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_counter_consistency_node_remove)
- [test_counter_consistency_node_remove](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9251/testReport/junit/counter_tests/TestCountersOnMultipleNodes/Tests___dtest___test_counter_consistency_node_remove_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_counter_consistency_node_remove)
- [test_counter_consistency_node_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9251/testReport/junit/counter_tests/TestCountersOnMultipleNodes/Tests___dtest___test_counter_consistency_node_decommission) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_counter_consistency_node_decommission)
- [test_counter_consistency_node_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9251/testReport/junit/counter_tests/TestCountersOnMultipleNodes/Tests___dtest___test_counter_consistency_node_decommission_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_counter_consistency_node_decommission)
- [test_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9251/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_decommission) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission)
- [test_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9251/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_decommission_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission)
- [test_cluster_cleanup_no_resurrection](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9251/testReport/junit/cleanup_test/TestCleanup/Tests___dtest___test_cluster_cleanup_no_resurrection) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_cleanup_no_resurrection)
- [test_cluster_cleanup_no_resurrection](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9251/testReport/junit/cleanup_test/TestCleanup/Tests___dtest___test_cluster_cleanup_no_resurrection_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_cleanup_no_resurrection)
- [test_hintedhandoff_removenode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9251/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_removenode) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_removenode)
- [test_hintedhandoff_removenode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9251/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_removenode_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_removenode)
#### Build Details:
- Duration: 6 hr 16 min
- Builder: spider8.cloudius-systems.com

piodul commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9251/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9251/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9785/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1761/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9251/)

@dawmd please look into the failures

dawmd commented:
Version 4:
* I didn't verify if the returned iterator was valid, which led to segmentation faults. Fixed

piodul commented:
I changed the title to be more descriptive

piodul commented:
Ugh, the CI job didn't run properly for some reason... retriggering

piodul commented:
> 19:58:02  Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: e361bc99-c69c-49de-9037-d8a607607687
> 19:58:02  java.util.NoSuchElementException

Rerunning once more here: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9414/

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9428/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9428/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9886/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1863/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9428/)
#### Build Details:
- Duration: 5 hr 59 min
- Builder: i-067754b1159a9c9ca (m5ad.8xlarge)

piodul commented:
Queued (01ff8108c1b08bc9d15d4ebc0cc989b89baf15ee)
